### PR TITLE
Implement multiple users with policies

### PR DESCRIPTION
**Pull request by ntk148v**
Closes: #94

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -20,85 +20,137 @@ import (
...
-
+// Register registers the API handlers under their correct routes
+// in the given router.
+func (a *API) Register(r *mux.Router) {
+	mw := middleware.New(log.With(a.logger, "component", "transport middleware"))
+	// General middleware
+	r.Use(mw.Instrument, mw.Logging, mw.RestrictDomain, mw.HandleCors)
+	// Unrestricted subrouter
+	pubr := r.PathPrefix("/public").Subrouter()
+	pubr.Use(mw.Instrument, mw.Logging, mw.RestrictDomain, mw.HandleCors)
```
Comment (vtdat) [inline api/api.go:100]: Reapply middleware to subrouter? is it a bit redundant

File: api/policy.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+	}
+	// Add new policy to Etcd
+	if _, err := a.policyEngine.AddPolicies(rules); err != nil {
+		a.respondError(w, apiError{
+			code: http.StatusInternalServerError,
+			err:  err,
+		})
+		return
+	}
+	if err := a.policyEngine.SavePolicy(); err != nil {
```
Comment (vtdat) [inline api/policy.go:72]: this method expects 1 params afaik https://github.com/ntk148v/etcd-adapter/blob/master/adapter.go#L141

File: api/policy.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	}
+	// Remove policy from Etcd
+	if _, err := a.policyEngine.RemovePolicies(rules); err != nil {
+		a.respondError(w, apiError{
+			code: http.StatusInternalServerError,
+			err:  err,
+		})
+		return
+	}
+	if err := a.policyEngine.SavePolicy(); err != nil {
```
Comment (vtdat) [inline api/policy.go:126]: ditto

File: api/policy.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+			code: http.StatusBadRequest,
+			err:  err,
+		})
+		return
+	}
+	for _, p := range pols {
+		rules = append(rules, []string{user, p.Path, p.Method})
+	}
+	// Remove policy from Etcd
+	if _, err := a.policyEngine.RemovePolicies(rules); err != nil {
```
Comment (vtdat) [inline api/policy.go:119]: this expects 3 params https://github.com/ntk148v/etcd-adapter/blob/master/adapter.go#L261

File: api/policy.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+			code: http.StatusBadRequest,
+			err:  err,
+		})
+		return
+	}
+	for _, p := range pols {
+		rules = append(rules, []string{user, p.Path, p.Method})
+	}
+	// Add new policy to Etcd
+	if _, err := a.policyEngine.AddPolicies(rules); err != nil {
```
Comment (vtdat) [inline api/policy.go:65]: this expects 3 params https://github.com/vCloud-DFTBA/faythe/pull/100/files#diff-fe22edb9ebb33fd9681f6a5df0a52cb5R240

File: api/api.go (hunk 1 above)
Comment (ntk148v) [reply api/api.go]: Nice catch.

File: api/token.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+	if err != nil {
+		a.respondError(w, apiError{
+			code: http.StatusInternalServerError,
+			err:  err,
+		})
+		return
+	}
+	if len(resp.Kvs) == 0 {
+		a.respondError(w, apiError{
+			code: http.StatusBadRequest,
```
Comment (vtdat) [inline api/token.go:48]: Should be `Unauthorized`

File: api/users.go [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+	_, err = a.etcdcli.DoPut(path, string(r))
+	if err != nil {
+		a.respondError(w, apiError{
+			code: http.StatusInternalServerError,
+			err:  errors.Wrap(err, "Unable to put a key-value pair into etcd"),
+		})
+		return
+	}
+	a.respondSuccess(w, http.StatusOK, nil)
+	return
```
Comment (vtdat) [inline api/users.go:132]: should we add baseline policy for user?

Comment (ntk148v) [reply api/users.go:132]: I thought about it before, but our objects contain very sensitive information like password, scale url... therefore, it's hard to choose a baseline policy.

File: api/token.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+		if err != nil {
+			a.respondError(w, apiError{
+				code: http.StatusInternalServerError,
+				err:  errors.Wrap(err, "Something went wrong"),
+			})
+			return
+		}
+		w.Header().Set("Content-Type", "application/json")
+		w.Header().Add("Authorization", fmt.Sprintf(bearerFormat, tokenString))
+		a.respondSuccess(w, http.StatusOK, "")
```
Comment (vtdat) [inline api/token.go:71]:
```suggestion
		a.respondSuccess(w, http.StatusOK, nil)
```

File: config/config.go [hunk 9]
```diff
@@ -115,27 +140,25 @@ type EtcdConfig struct {
...
 
 const (
-	etcdDefaultDialTimeout      = 5 * time.Second
-	etcdDefaultKeepAliveTime    = 5 * time.Second
-	etcdDefaultKeepAliveTimeOut = 6 * time.Second
+	etcdDefaultDialTimeout       = 5 * time.Second
+	etcdDefaultKeepAliveTime     = 5 * time.Second
+	etcdDefaultKeepAliveTimeOut  = 6 * time.Second
+	jwtDefaultSigningMethod      = "RS256"
+	jwtDefaultTTL                = 15 * time.Minute
```
Comment (vtdat) [inline config/config.go:147]: raise this default value please. 15m is way too small

File: api/users.go (hunk 7 above)
Comment (vtdat) [reply api/users.go:132]: like if a person doesn't get `clouds` view pocily, how did him choose clouds to create scalers, healers, silencers?

File: config/config.go (hunk 9 above)
Comment (ntk148v) [reply config/config.go]: OK, increase to 60m.

File: api/users.go (hunk 7 above)
Comment (ntk148v) [reply api/users.go]: Ok

**Comment by vtdat** [comment]
https://github.com/vCloud-DFTBA/faythe/pull/100/files#diff-a39c9f7aa728d5fa3b973bc6ba49228aR51

> etcdctl --username

**Comment by vtdat** [comment]
Since you're using external lib for Authentication middleware, the response does not match our pattern


```bash
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8600 (#0)
> POST /users HTTP/1.1
> Host: localhost:8600
> User-Agent: curl/7.58.0
> Accept: */*
> Authorization: Bearer 
> Content-Length: 246
> Content-Type: multipart/form-data; boundary=------------------------002c5a2e485c9853
> 
< HTTP/1.1 401 Unauthorized
< Access-Control-Allow-Credentials: true
< Access-Control-Allow-Headers: Accept, Authorization, Content-Type, Origin, X-Request-With
< Access-Control-Allow-Methods: GET, POST, DELETE, PUT
< Access-Control-Allow-Origin: 
< Access-Control-Expose-Headers: Date
< Cache-Control: no-cache, no-store, must-revalidate
< Content-Type: text/plain; charset=utf-8
< X-Content-Type-Options: nosniff
< Date: Fri, 12 Jun 2020 02:52:40 GMT
< Content-Length: 31
* HTTP error before end of send, stop sending
< 
crypto/rsa: verification error
* Closing connection 0
```

**Pull request merged**